# SQL Part 1 Studio

### Let's practice your SQL quering skills!  For each question, work along in the notebook adding your query and answering the questions. 

### **DataSet**

### <span style="font-size: 14px; font-family: -apple-system, BlinkMacSystemFont, sans-serif;">We will be using the Goodbooks-10k dataset from the exercises in the prep work.&nbsp; Feel free to reference your notebook.</span>

- The dataset can be found here: [goodbooks-10k](https://www.kaggle.com/zygmunt/goodbooks-10k)
- You can access the **BooksDB** in the Launchcode server.

### **Business Issue**

You work for a small independant book store and you want to increase sales by running various promotions.  Start by becoming more familar with the BooksDB.

## Part 1. BooksDB Questions

<span style="color: rgba(0, 0, 0, 0.87); font-family: -apple-system, system-ui, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; background-color: rgb(255, 255, 255);">Question 1:&nbsp; Write a query of the books table, returning the TOP 100 results and&nbsp; includes, book_id, authors, title, average_rating.&nbsp; Use an alias for at least one column and in descending order of rating.&nbsp; What is the number one book?</span>

In [1]:
-- return top 100 results
SELECT TOP 100 book_id, authors, title, average_rating AS RATING_AVG
FROM Books 
ORDER BY RATING_AVG DESC

-- Book "The Complete Calvin and Hobbes" by Bill Watterson is the number one book.

(100 rows affected)

Total execution time: 00:00:00.120

book_id authors title RATING_AVG 24812 Bill Watterson The Complete Calvin and Hobbes 4.82 17332218 Brandon Sanderson Words of Radiance (The Stormlight Archive, #2) 4.77 8 J.K. Rowling, Mary GrandPré Harry Potter Boxed Set, Books 1-5 (Harry Potter, #1-5) 4.77 5031805 Anonymous, Lane T. Dennis, Wayne A. Grudem ESV Study Bible 4.76 95602 Francine Rivers Mark of the Lion Trilogy 4.76 24814 Bill Watterson It's a Magical World: A Calvin and Hobbes Collection 4.75 862041 J.K. Rowling Harry Potter Boxset (Harry Potter, #1-7) 4.74 70489 Bill Watterson There's Treasure Everywhere: A Calvin and Hobbes Collection 4.74 10 J.K. Rowling Harry Potter Collection (Harry Potter, #1-6) 4.73 59715 Bill Watterson The Authoritative Calvin and Hobbes: A Calvin and Hobbes Treasury 4.73 24815 Bill Watterson The Indispensable Calvin and Hobbes 4.73 17927395 Sarah J. Maas A Court of Mist and Fury (A Court of Thorns and Roses, #2) 4.72 70487 Bill Watterson Attack of the Deranged Mutant Killer Monster Snow Goons 4.72 24816 Bill Watterson Homicidal Psycho Jungle Cat: A Calvin and Hobbes Collection 4.71 121792 Bill Watterson The Revenge of the Baby-Sat 4.71 2350129 The Church of Jesus Christ of Latter-day Saints Preach My Gospel: A Guide To Missionary Service 4.71 24818 Bill Watterson The Days Are Just Packed: A Calvin and Hobbes Collection 4.68 265205 Anonymous, Ronald A. Beers, Ronald A. Beers Life Application Study Bible: NIV 4.67 9329354 Brandon Sanderson The Way of Kings, Part 1 (The Stormlight Archive #1.1) 4.67 99298 J.K. Rowling, Mary GrandPré The Harry Potter Collection 1-4 (Harry Potter, #1-4) 4.66 5883777 Anonymous The Holy Bible: English Standard Version 4.66 24494 Bill Watterson The Calvin and Hobbes Lazy Sunday Book 4.66 43070 Bill Watterson The Essential Calvin and Hobbes: A Calvin and Hobbes Treasury 4.65 18337259 Daniel Abraham, George R.R. Martin, Tommy Patterson A Game of Thrones: Comic Book, Issue 1 4.65 23753 Neil Gaiman, Mike Dringenberg, Chris Bachalo, Michael Zulli, Kelly Jones, Charles Vess, Colleen Doran, Malcolm Jones III, Steve Parkhouse, Daniel Vozzo, Lee Loughridge, Steve Oliff, Todd Klein, Dave McKean, Sam Kieth The Absolute Sandman, Volume One 4.65 7235533 Brandon Sanderson The Way of Kings (The Stormlight Archive, #1) 4.64 24813 Bill Watterson The Calvin and Hobbes Tenth Anniversary Book 4.63 9814682 George R.R. Martin A Song of Ice and Fire (A Song of Ice and Fire, #1-4) 4.63 12177850 George R.R. Martin A Song of Ice and Fire (A Song of Ice and Fire, #1-5) 4.63 481749 James E. Talmage Jesus the Christ 4.63 46292 Hafez The Divan 4.63 6801614 Eiichirō Oda One Piece, Volume 38: Rocketman!! (One Piece, #38) 4.63 22299763 Leigh Bardugo Crooked Kingdom (Six of Crows, #2) 4.62 32075671 Angie Thomas The Hate U Give 4.62 136251 J.K. Rowling, Mary GrandPré Harry Potter and the Deathly Hallows (Harry Potter, #7) 4.61 77727 Bill Watterson, G.B. Trudeau Calvin and Hobbes 4.61 280111 Anonymous Holy Bible: New International Version 4.61 2223324 John Williams Harry Potter and the Chamber of Secrets: Sheet Music for Flute with C.D 4.61 6339989 Richelle Mead Vampire Academy Collection (Vampire Academy, #1-3) 4.61 596259 Gordon B. Hinckley Standing for Something: 10 Neglected Virtues That Will Heal Our Hearts and Homes 4.61 47697 Alisa Kwitney, Neil Gaiman The Sandman: King of Dreams 4.61 18006496 Sarah J. Maas Queen of Shadows (Throne of Glass, #4) 4.6 95617 Francine Rivers, Richard Ferrone A Voice in the Wind (Mark of the Lion, #1) 4.6 46601 Francine Rivers An Echo in the Darkness (Mark of the Lion, #2) 4.6 7025086 Christopher Yost, Pasqual Ferry, Orson Scott Card Ender's Game, Volume 2: Command School 4.6 18335634 Cassandra Clare Clockwork Princess (The Infernal Devices, #3) 4.59 30 J.R.R. Tolkien J.R.R. Tolkien 4-Book Boxed Set: The Hobbit and The Lord of the Rings 4.59 2299110 Sherrilyn Kenyon Acheron (Dark-Hunter #14) 4.59 119 Gary Russell The Lord of the Rings: The Art of The Fellowship of the Ring 4.59 71252 Neil Gaiman, Marc Hempel, R

Question 2: Write a query to find the least popular book.

In [2]:
-- return least popular book
select * from Books
where average_rating=(select min(average_rating) from Books);

(1 row affected)

Total execution time: 00:00:00.055

id book_id best_book_id work_id books_count isbn isbn13 authors original_publication_year original_title title language_code average_rating ratings_count work_ratings_count work_text_reviews_count ratings_1 ratings_2 ratings_3 ratings_4 ratings_5 image_url small_image_url 1793 105578 105578 101770 25 345498321 9780345498330 Chetan Bhagat 2005 One Night @ The Call Center One Night at the Call Center eng 2.47 40718 44833 1244 10559 14125 11872 5177 3100 https://images.gr-assets.com/books/1320500924m/105578.jpg https://images.gr-assets.com/books/1320500924s/105578.jpg

Question 3: Which tag is the most popular?

In [3]:
-- return most popular tag
select * from Books
where average_rating =(select max(average_rating) from Books);

(1 row affected)

Total execution time: 00:00:00.055

id book_id best_book_id work_id books_count isbn isbn13 authors original_publication_year original_title title language_code average_rating ratings_count work_ratings_count work_text_reviews_count ratings_1 ratings_2 ratings_3 ratings_4 ratings_5 image_url small_image_url 3628 24812 24812 25599 14 740748475 9780740748480 Bill Watterson 2005 The Complete Calvin and Hobbes The Complete Calvin and Hobbes eng 4.82 28900 29968 861 120 154 693 3117 25884 https://images.gr-assets.com/books/1473064526m/24812.jpg https://images.gr-assets.com/books/1473064526s/24812.jpg

Question 4: What is the name of the most popular tag?

In [4]:
-- return name of most popular tag
select title AS "Most Popular Tag" from Books
where average_rating=(select max(average_rating) from Books);

(1 row affected)

Total execution time: 00:00:00.051

Most Popular Tag The Complete Calvin and Hobbes

Question 5: How many books where released in the first decade of 2000?

In [5]:
-- return number of books published in the first decade of 2000
SELECT COUNT(books_count) FROM Books where original_publication_year BETWEEN 2000 and 2009

(1 row affected)

Total execution time: 00:00:00.052

(No column name) 3121

Question 6: How many book titles contain the word Happy?

In [6]:
-- return number of books that contain the word Happy
SELECT count(books_count) 
  FROM Books
  where title like '%Happy%'

(1 row affected)

Total execution time: 00:00:00.064

(No column name) 13

Question 7: List the books from the top 3 authors from question 1.  If there is more than one author just use the first one. Sort the titles <span style="background-color: rgba(127, 127, 127, 0.1); font-family: -apple-system, BlinkMacSystemFont, sans-serif;">alphabetically by author and then by average_rating, best rated to lowest. Does order matter in sorting?</span>

In [7]:
-- return all books from the top 3 authors, sort by author and rating
SELECT TOP 3 book_id, authors, title, average_rating AS RATING_AVG
FROM Books 
ORDER BY RATING_AVG DESC

SELECT authors, title, average_rating from Books
where authors = 'Bill Watterson'
ORDER BY title, average_rating desc

SELECT authors, title, average_rating from Books
where authors = 'J.K. Rowling, Mary GrandPré'
ORDER BY title, average_rating desc

(3 rows affected)

(12 rows affected)

(8 rows affected)

Total execution time: 00:00:00.111

book_id authors title RATING_AVG 24812 Bill Watterson The Complete Calvin and Hobbes 4.82 17332218 Brandon Sanderson Words of Radiance (The Stormlight Archive, #2) 4.77 8 J.K. Rowling, Mary GrandPré Harry Potter Boxed Set, Books 1-5 (Harry Potter, #1-5) 4.77

authors title average_rating Bill Watterson Attack of the Deranged Mutant Killer Monster Snow Goons 4.72 Bill Watterson Homicidal Psycho Jungle Cat: A Calvin and Hobbes Collection 4.71 Bill Watterson It's a Magical World: A Calvin and Hobbes Collection 4.75 Bill Watterson The Authoritative Calvin and Hobbes: A Calvin and Hobbes Treasury 4.73 Bill Watterson The Calvin and Hobbes Lazy Sunday Book 4.66 Bill Watterson The Calvin and Hobbes Tenth Anniversary Book 4.63 Bill Watterson The Complete Calvin and Hobbes 4.82 Bill Watterson The Days Are Just Packed: A Calvin and Hobbes Collection 4.68 Bill Watterson The Essential Calvin and Hobbes: A Calvin and Hobbes Treasury 4.65 Bill Watterson The Indispensable Calvin and Hobbes 4.73 Bill Watterson The Revenge of the Baby-Sat 4.71 Bill Watterson There's Treasure Everywhere: A Calvin and Hobbes Collection 4.74

authors title average_rating J.K. Rowling, Mary GrandPré Harry Potter and the Chamber of Secrets (Harry Potter, #2) 4.37 J.K. Rowling, Mary GrandPré Harry Potter and the Deathly Hallows (Harry Potter, #7) 4.61 J.K. Rowling, Mary GrandPré Harry Potter and the Goblet of Fire (Harry Potter, #4) 4.53 J.K. Rowling, Mary GrandPré Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 J.K. Rowling, Mary GrandPré Harry Potter and the Order of the Phoenix (Harry Potter, #5) 4.46 J.K. Rowling, Mary GrandPré Harry Potter and the Sorcerer's Stone (Harry Potter, #1) 4.44 J.K. Rowling, Mary GrandPré Harry Potter Boxed Set, Books 1-5 (Harry Potter, #1-5) 4.77 J.K. Rowling, Mary GrandPré The Harry Potter Collection 1-4 (Harry Potter, #1-4) 4.66

Question 8: Take a close look at J.K. Rowling.  Is her top rated book on this list from question 1?  Why might this be? How could we account for this?

> <span style="color: rgb(0, 0, 0); font-family: &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; background-color: rgb(255, 255, 255);">Yes J.K. Rowling's "Harry Potter Boxed Set, Books 1-5 (Harry Potter, #1-5)" book is listed in 3rd place as it had rating average of "4.77".</span>

Try updating your above query to include all of J.K. Rowling's books.

Are there any other issues?

No Issues

In [8]:
-- Try updating your above query to include all of J.K. Rowling's books
SELECT authors, title, average_rating from Books
where authors = 'J.K. Rowling, Mary GrandPré'
ORDER BY title, average_rating desc

(8 rows affected)

Total execution time: 00:00:00.099

authors title average_rating J.K. Rowling, Mary GrandPré Harry Potter and the Chamber of Secrets (Harry Potter, #2) 4.37 J.K. Rowling, Mary GrandPré Harry Potter and the Deathly Hallows (Harry Potter, #7) 4.61 J.K. Rowling, Mary GrandPré Harry Potter and the Goblet of Fire (Harry Potter, #4) 4.53 J.K. Rowling, Mary GrandPré Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 J.K. Rowling, Mary GrandPré Harry Potter and the Order of the Phoenix (Harry Potter, #5) 4.46 J.K. Rowling, Mary GrandPré Harry Potter and the Sorcerer's Stone (Harry Potter, #1) 4.44 J.K. Rowling, Mary GrandPré Harry Potter Boxed Set, Books 1-5 (Harry Potter, #1-5) 4.77 J.K. Rowling, Mary GrandPré The Harry Potter Collection 1-4 (Harry Potter, #1-4) 4.66

Question 9: Write a query that returns the number of Authors whose first name is between rock and roll.

In [9]:
-- return number of authors whose first name is between rock and roll
SELECT count(authors) from Books
where authors between 'rock' and 'roll'



(1 row affected)

Total execution time: 00:00:00.091

(No column name) 25

## Part 2. Write your own question and write a query to answer it.

Your Question:  Click here to add

In [10]:
-- Add your querry here
select authors, books_count from Books where books_count=(select max(books_count) from Books)

(1 row affected)

Total execution time: 00:00:00.184

authors books_count Jane Austen 3455

Your Question: Click here to add

In [11]:
-- Add your querry here
select top 10 book_id, authors, title from Books where original_publication_year = '1992'

(10 rows affected)

Total execution time: 00:00:00.047

book_id authors title 32499 John Grisham The Pelican Brief 5203 Wally Lamb She's Come Undone 830 Neal Stephenson Snow Crash 5364 Diana Gabaldon Dragonfly in Amber (Outlander, #2) 29044 Donna Tartt The Secret History 9539 Robert Jordan The Shadow Rising (Wheel of Time, #4) 32508 Michael Connelly The Black Echo (Harry Bosch, #1; Harry Bosch Universe, #1) 32692 Stephen King Gerald's Game 10625 Stephen King, Dominique Dill Dolores Claiborne 11713 Michael Ondaatje The English Patient

**Some Question Examples:**

Write a query that returns the longest title?

Which Author has written the most books?  

Top  ten books written the year I was born?